infra: add docs for hardware keys (#1074)
rust-lang-nursery/rust-forge: src/infra/docs/hardware-security-keys.md

SSH keys hardware-backed makes private SSH keys effortlessly portable across
different machines.

For example, if you want to use a Yubikey-backed SSH key with your Github account,
[OpenSSH built-in support for FIDO2 authentication] may be the easiest way to get
started.

The setup will differ depending on your operating system, but keep in mind that
you may also need to tweak your Git configuration to make sure that operations
like commit signing continue to work as expected. In particular, we recommend
not prompting the passkey PIN for every Git operation, since it might be
counterproductive.

As an example, this command will define a new resident SSH key through FIDO2
and won't prompt your passkey everytime, nor will it require touching the
hardware key for Git operations.

```shell
ssh-keygen -t ed25519-sk \
  -O resident \
  -O application=ssh:git \
  -C "me@email.com"
```

Hardware-backed SSH keys are optional for now, but might be required for certain
administrative infrastructure tasks in the future. You can check which resident
SSH keys are available in your YubiKey either with CLI or Yubico Authenticator
app.

```shell

> uvx --from yubikey-manager ykman fido credentials list

Reviewer: ```suggestion
$ uvx --from yubikey-manager ykman fido credentials list
```
this make it clear that the rest is stdout

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-forge/1074/9a6e153dabe56c5938ad8bae996618961a66ec27..e7ebe9c1ad717bd4e15c60e19032735aaba79ef5)*

### Supported firmware versions

Based on existing [security advisories], only YubiKeys with firmware
**version v5.7.4** or newer are allowed. You can use either the `ykman`
CLI or Yubico Authenticator to [check the current firmware version] of a YubiKey.

### Recommended first steps

There are at least two official open-source tools provided by Yubico tools that
help setting up an Yubikey:

* [Yubico Authenticator] (sources : [Desktop and Android] | [iOS]):  
  * The Desktop app is pretty capable of handling most of the tasks.  
  * You can download the Mobile App directly on Google Play or Apple App Store.  
* Yubico Manager CLI (sources: [yubikey-manager]): for advanced use cases.

As a good first step, the Rust infrastructure team recommends using the Yubico
Authenticator Desktop app to change the default values for:

* The [Passkeys PIN default value reference]
* The [PIV app PIN and PUK default values reference]

For those who prefer the CLI, [uvx] might be an alternative to run the `ykman`
Python utility without installing it globally in your system or managing Python
installations. You can refer to the [ykman online documentation] to learn more
about subcommands.

```shell

> ➜  uvx --from yubikey-manager ykman --help

Reviewer: ```suggestion
uvx --from yubikey-manager ykman --help
```
now that there is no stdout anymore, there's no need to write the prompt character.

Author: Changed [here](https://github.com/rust-lang/rust-forge/pull/1074/changes/e7ebe9c1ad717bd4e15c60e19032735aaba79ef5)

# Hardware Security Keys in the Rust Project

## Overview

### Security hardware keys in Rust Infrastructure

> Hardware security keys drastically improve security by providing unphishable

Reviewer 2: nit: either "Security hardware keys" or "Hardware security keys" :slightly_smiling_face: 

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-forge/1074/9a6e153dabe56c5938ad8bae996618961a66ec27..33e72f07007181a4134ec973377761bfb527e189)*

Author: Favored the later [here](https://github.com/rust-lang/rust-forge/pull/1074/changes/0ac72df94f9df3b0d45272625944b7d1a51e09f6). Thanks for the catch!

> Hardware security keys drastically improve security by providing unphishable

Reviewer: ```suggestion
Hardware security keys improve security by providing unphishable
```

Author: Added [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

> protection for sensitive code and infrastructure.The Rust infrastructure team

Reviewer: ```suggestion
protection for sensitive code and infrastructure. The Rust infrastructure team
```
what do you mean with sensitive code?

Author: I wrote _code_, but after a second thought, I meant _systems_ actually. Rephrased [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/fb97845d696faad9014cdf61d9d97f10907adc8c), but happy to make this shorter as well, lmk

officially supports hardware security keys in partnership with
[Yubico Secure it Forward] program.

The Rust Foundation provides YubiKeys to Rust Project members who need access
to critical infrastructure systems. If you are eligible for such a grant and
would like to get the recommended YubiKeys for free, get in touch with the

> [T-infra in Zulip].

Reviewer: I would move the "if you are eligible" part after the list below.

Author: Changed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b) as you suggested!

Currently, the members of the following teams are eligible for this grant:

* `infra`  
* `crates.io`
* `docs.rs`  
* `release`
* `triagebot`
* `bors`

## YubiKeys support in the Rust infrastructure

### Supported models

The Rust infrastructure team has validated the [Yubico Series 5 USB/NFC models]

> products and will officially support Yubico Series 5 keys for any issues a

Reviewer: ```suggestion
products and officially supports the Yubico Series 5 keys for any issues a
```

Author: Changed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

Project member might have.

### Supported firmware versions

Based on existing [security advisories], only YubiKeys with firmware

> **version v5.7.4** or newer are fully supported. You can use either the `ykman`

Reviewer: ```suggestion
**version v5.7.4** or newer are allowed. You can use either the `ykman`
```

Author: Changed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

CLI or Yubico Authenticator to [check the current firmware version] of a YubiKey.

### Recommended first steps

There are at least two official open-source tools provided by Yubico tools that
help setting up an Yubikey:

* [Yubico Authenticator] (sources : [Desktop and Android] | [iOS]):  
  * The Desktop app is pretty capable of handling most of the tasks.  
  * You can download the Mobile App directly on Google Play or Apple App Store.  

> * Yubico Manager CLI (sources: [yubikey-manager]): for advanced use cases

Reviewer: ```suggestion
* Yubico Manager CLI (sources: [yubikey-manager]): for advanced use cases.
```

Author: Added [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

As a good first step, the Rust infrastructure team recommends using the Yubico

> Authenticator Desktop app to change the default values for :

Reviewer: ```suggestion
Authenticator Desktop app to change the default values for:
```

Author: Fixed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

* The [Passkeys PIN default value reference]
* The [PIV app PIN and PUK default values reference]

For those who prefer the CLI, [uvx] might be an alternative to run the `ykman`
Python utility without installing it globally in your system or managing Python
installations. You can refer to the [ykman online documentation] to learn more
about subcommands.

```shell
➜  uvx --from yubikey-manager ykman --help
Installed 15 packages in 15ms
Usage: ykman [OPTIONS] COMMAND [ARGS]...

  Configure your YubiKey via the command line.

  Examples:

    List connected YubiKeys, only output serial number:

>     $ ykman list --serials
> 
>     Show information about YubiKey with serial number 123456:
>     $ ykman --device 123456 info
> 
> ```

Reviewer: ```suggestion
installations:

```sh
uvx --from yubikey-manager ykman --help
```

You can refer to the [ykman online documentation] to learn more
about subcommands.
```
Imo the stdout of the help menu is not helpful so we could remove it.

Author: Changed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b), left only the command

[official steps]. At some point, you'll be prompted with the following dialog:

![yubikeys gh setup](img/yubikey-gh-setup.jpg)

The setup flow will vary depending on the system, and might also change
according to web browser extensions you have installed (e.g. 1password,
which offers itself as an option during such a setup) but in general you
should mind the *security key* web browser dialog.

This step triggers the check for the human presence against the hardware
key. Touch your YubiKey to confirm the authentication.

Our [upcoming infrastructure policy] mandates using this option for all
compatible critical Rust Infrastructure systems.

### Hardware-backed multi factor authentication

For systems that require multi-factor authentication but only support TOTP
codes as a secure MFA method, YubiKeys can still work as an alternative to
existing solutions like Google Authenticator. In this case, you can use the
Yubico Authenticator app to set up [hardware-backed TOTP code generation].

This option provides a way to have working TOTP codes in both a mobile phone
and a desktop system without relying on third-party cloud systems, but also

> has the con of coupling TOTP code generation with a physical device : similarly

Reviewer: ```suggestion
has the con of coupling TOTP code generation with a physical device: similarly
```

Author: Fixed [here](https://github.com/rust-lang/rust-forge/pull/1074/commits/c1d3302a84e45d5df5c95e69f66e0c8c77a4a14b)

to using an offline-only TOTP authenticator app, losing the Yubikey means
losing access to TOTP codes, which requires additional diligence regarding
backing up recovery codes.

> Note that setting up hardware-backed TOTP codes is totally optional for

Reviewer: ```suggestion
Note that setting up hardware-backed TOTP codes is optional for
```

Rust Project members.

### Hardware-backed SSH keys

For those who want additional security for SSH authentication, YubiKeys
offer different [options for hardware-backed SSH key pairs]. Having your
SSH keys hardware-backed makes private SSH keys effortlessly portable across
different machines.

For example, if you want to use a Yubikey-backed SSH key with your Github account,
[OpenSSH built-in support for FIDO2 authentication] may be the easiest way to get
started.

The setup will differ depending on your operating system, but keep in mind that
you may also need to tweak your Git configuration to make sure that operations
like commit signing continue to work as expected. In particular, we recommend
not prompting the passkey PIN for every Git operation, since it might be
counterproductive.

As an example, this command will define a new resident SSH key through FIDO2
and won't prompt your passkey everytime, nor will it require touching the

> hardware key for Git operations.

Reviewer: so it only requires the hardware key being connected? Is this what we decided at the all hands?

I think we should suggest touching the hardware key, or at least explain the risk connected to not having to touch the hardware key. Which imo are that if someone has a remote access to your terminal they can run git operations.

Author: > so it only requires the hardware key being connected?

Yes, in this case.

> Is this what we decided at the all hands?

The decisions we made in the All-hands were mostly around using PIV, which is another option to have hardware-backed SSH keys. I've mentioned FIDO2 as an example _mostly_ to illustrate this capability, but I do want to add more on PIV as a follow-up of https://github.com/rust-lang/team/issues/2501

Happy to remove this section for now, lmk :)

Reviewer: I'm fine leaving this section. Probably it's safest to suggest touching the yubikey.
I leave the final decision to you!

Reviewer 2: for what it's worth, touching the FIDO2 token is what I am routinely asked when I pay online and it "feels" the expected action to be prompted.

I am not too deep into the implications or threat scenarios but if there are no opposing opinions (or this specific topic was just not discussed), could we remove `-O no-touch-required ` and leave as default to touch the token?

Author: Took this suggestion and improved the example [here](https://github.com/rust-lang/rust-forge/pull/1074/changes/e7ebe9c1ad717bd4e15c60e19032735aaba79ef5) :)

SSH keys are available in your YubiKey either with CLI or Yubico Authenticator
app.

```shell
➜  uvx --from yubikey-manager ykman fido credentials list
Enter your PIN:
Credential ID  RP ID    Username  Display name
86707903...    ssh:git  openssh   openssh

```

![yubikeys ssh setup](img/yubikey-sshkey.jpg)

### PIV and attestations

YubiKeys are compatible with [Personal Identity Verification (PIV)] for smart
cards, which allows using them for encryption and signing operations on top
of this particular standard.

Some capabilities backed by PIV will be introduced later in the Rust Project.
For now, you may want to [watch this issue] to follow-up on this topic.

## FAQ

### Which services should I use 2FA with my YubiKey?

Please check our [upcoming infrastructure policy].

### How do I handle my preferred cloud CLI authentication with my YubiKey?

For Gcloud CLI, the [web-based authentication flow] will automatically prompt your
2FA method of choice, and after that, no additional steps are required. You just
need to configure it as your preferred [MFA method in your Google account].

For AWS CLI, you should stick with [AWS SSO user sessions] rather than IAM user

> sessions when setting up your CLI configuration. This flow willprompt your 2FA

Reviewer 2: ```suggestion
sessions when setting up your CLI configuration. This flow will prompt your 2FA
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-forge/1074/9a6e153dabe56c5938ad8bae996618961a66ec27..33e72f07007181a4134ec973377761bfb527e189)*

Author: Fixed [here](https://github.com/rust-lang/rust-forge/pull/1074/changes/0ac72df94f9df3b0d45272625944b7d1a51e09f6)
